Replay view with view_func instead of as_strided in meta_utils for NT

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112206
* __->__ #112205

Currently meta_utils relies on as_strided when handling the view case (recursively meta-ify the base, and then do as_strided to simulate the view), but NestedTensor does not support as_strided today (though maybe it could?), so what we want to do instead is call Tensor. _view_func. Conveniently,  _view_func IS always available for nested tensors.

A detail to note is that _view_func actually incurs a guard because it needs to perform some metadata checks to make sure the view is still valid. This PR adds Tensor._unsafe_view_func which can avoid that.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

soulitzer committed:
Support passing view of NT to compiled fn

[ghstack-poisoned]

soulitzer commented on torch/_subclasses/meta_utils.py:
```diff
@@ -382,7 +383,11 @@ def is_c_of_r(complex_dtype, real_dtype):
+                                        r = base.as_strided(sizes, strides, storage_offset)
+                                    else:
+                                        # nested tensor does not support as_strided
+                                        r = t._view_func_unsafe(base)
```
TODO: We can probably easily handle the other cases too

soulitzer replied:
More generally, I should call view_func_unsafe instead of as_strided for any tensor that does not support as_strided (but that attribute is not binded to python today)

ezyang replied:
Can you explain in the comments here why calling unsafe is OK and why not calling it would not work?

jbschlosser replied:
> TODO: We can probably easily handle the other cases too

I think we should do this :p or at least error nicely for NT on those branches

soulitzer replied:
Updated to actually handle these cases.

ezyang commented on test/dynamo/test_subclasses.py:
```diff
@@ -700,6 +700,21 @@ def fn(x):
+        # Check metadata and values are correct
+        self.assertTrue(out.size() == out_ref.size())
+        self.assertTrue(out.stride() == out_ref.stride())
+        self.assertTrue(torch.allclose(out.values(), out_ref.values()))
```
Can you test that no guards get incurred?

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -313,7 +313,7 @@ def sym_sizes_strides_storage_offset(t, src):
                         callback,
                         source=AttrSource(source, "_base"),
-                        dynamic_dims=base_dynamic_dims,
+                        dynamic_dims=None,
```
What's going on here?

jbschlosser replied:
This looks like a hack to get around the mismatch between `len(base_dynamic_dims)` and dims of the values and offsets that causes problems for NT (also see https://github.com/pytorch/pytorch/pull/112185#discussion_r1373715054).

For the purposes of this PR, this should definitely be removed.

soulitzer replied:
Whoops, fixed!

jbschlosser reviewed (commented):
Thanks for the quick fix!

soulitzer committed:
Update on "Replay view with view_func instead of as_strided in meta_utils for NT"


Currently meta_utils relies on as_strided when handling the view case (recursively meta-ify the base, and then do as_strided to simulate the view), but NestedTensor does not support as_strided today (though maybe it could?), so what we want to do instead is call Tensor. _view_func. Conveniently,  _view_func IS always available for nested tensors.

A detail to note is that _view_func actually incurs a guard because it needs to perform some metadata checks to make sure the view is still valid. This PR adds Tensor._unsafe_view_func which can avoid that.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
